### PR TITLE
Welcome screen facelift

### PR DESCRIPTION
**Pull request by zbettenbuk**
The PR is 2 commits:
- Generalize the safe area and use native components instead of custom header padding calculation (and replace the old components on the Settings screen)
- Add the facelifted Welcome page together with the new general components.

### CODE REVIEW

File: react/features/base/react/components/native/styles.js [hunk 1]
```diff
@@ -9,13 +9,27 @@ const HEADER_HEIGHT = 44;
...
+     * The topmost container of the animated side bar.
+     */
+    animatedMenuContainer: {
+        bottom: 0,
+        flexDirection: 'row',
+        left: -SIDEBAR_WIDTH,
+        position: 'absolute',
+        top: 0,
+        width: SIDEBAR_WIDTH,
+        zIndex: 999999
```
Comment (saghul) [inline react/features/base/react/components/native/styles.js]: why do we need such zindex? can we make sure this is the topmost component in the welcome page hierarchy instead?

File: react/features/welcome/components/WelcomePage.native.js [hunk 2]
```diff
@@ -65,49 +86,85 @@ class WelcomePage extends AbstractWelcomePage {
...
+     */
+    _onShowSideBar() {
+        this.props.dispatch(showSideBar());
+    }
+
+    /**
+     * Handles the audio-video switch changes.
+     *
+     * @private
+     * @param {boolean} startWithVideo - The ! of the new value.
```
Comment (saghul) [inline react/features/welcome/components/WelcomePage.native.js]: typo? the what of the new value?

File: react/features/welcome/components/WelcomePage.native.js [hunk 3]
```diff
@@ -153,34 +210,23 @@ class WelcomePage extends AbstractWelcomePage {
...
-                    { t('welcomepage.privacy') }
-                </Link>
-                <Link
-                    style = { styles.legaleseItem }
-                    url = { SEND_FEEDBACK_URL }>
-                    { t('welcomepage.sendFeedback') }
-                </Link>
-            </View>
-        );
+        dispatch(setAudioOnly(!show));
```
Comment (saghul) [inline react/features/welcome/components/WelcomePage.native.js]: This block doesn't look correct. If there is a video track, we shouldn't destroy it, just by muting it, which setAudioOnly already does should be enough.

File: react/features/welcome/reducer.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import { HIDE_SIDE_BAR, SHOW_SIDE_BAR } from './actionTypes';
+
+/**
+ * Reduces the Redux actions of the feature features/recording.
+ */
+ReducerRegistry.register('features/welcome', (state = {
+    sideBarVisible: false
+}, action) => {
+    switch (action.type) {
+    case HIDE_SIDE_BAR:
```
Comment (saghul) [inline react/features/welcome/reducer.js]: I suggest you fold these 2 actions into 1: SET_SIDEBAR_VISIBILITY, same as we do with the toolbox or fimlstrip.

**Comment by saghul** [comment]
img_0023

There are a number of problems in a non-X iPhone:

- status bar is now dark? (text is black)
- top space is not accounted for

Side comment: me no likey the new join button color :-)

**Comment by saghul** [comment]
screenshot_20180206-165722

There is a border in the view which looks bad. I think it might be related to this: https://github.com/jitsi/jitsi-meet/commit/2094b15432383c00380b5c471b3764651255d927 because we use it here: https://github.com/jitsi/jitsi-meet/blob/master/react/features/welcome/components/styles.js#L5

File: react/features/welcome/components/WelcomePage.native.js (hunk 2 above)
Comment (zbettenbuk) [reply react/features/welcome/components/WelcomePage.native.js]: Negate :)

File: react/features/base/react/components/native/styles.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import {
+    BoxModel,
+    ColorPalette,
+    createStyleSheet
+} from '../../../styles';
+
+const HEADER_COLOR = ColorPalette.blue;
+const HEADER_HEIGHT = 44;
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/base/react/components/native/styles.js]: it's from the iOS design guidelines

File: react/features/base/react/components/native/styles.js [hunk 6]
```diff
@@ -9,13 +9,27 @@ const HEADER_HEIGHT = 44;
...
 /**
  * The styles of the React {@code Components} of the generic components
  * in the app.
  */
 export default createStyleSheet({
 
+    /**
+     * The topmost container of the animated side bar.
+     */
+    animatedMenuContainer: {
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/base/react/components/native/styles.js]: renamed it to sideMenuContainer, because container is too generic in this context

File: react/features/base/styles/components/styles/PlatformElements.native.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * Platform specific header button (e.g. back, menu...etc).
+     */
+    headerButton: {
+        alignSelf: 'center',
+        color: ColorPalette.white,
+        fontSize: 26,
+        paddingRight: 22,
+        zIndex: 9999
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/base/styles/components/styles/PlatformElements.native.js]: it was, when I had the app-title, but now that we only have a switch, it's not needed anymore. Thanks!

File: react/features/base/react/components/native/SideBar.js [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    /**
+     * Sets the side menu visible or hidden.
+     *
+     * @private
+     * @param {boolean} show - The new expected visibility value.
+     * @returns {void}
+     */
+    _setShow(show) {
+        if (this.state.showSideBar !== show) {
+            if (show) {
```
[the remark this answers is not in the document]
Comment (zbettenbuk) [reply react/features/base/react/components/native/SideBar.js]: we discussed it offline

File: react/features/analytics/AnalyticsEvents.js [hunk 9]
```diff
@@ -97,6 +97,19 @@ export const createAudioOnlyDisableEvent = function() {
...
 };
 
+/**
+ * Creates an event which indicates that the audio-only mode has been turned
+ * on.
+ *
+ * @returns {Object} The event in a format suitable for sending via
+ * sendAnalytics.
+ */
+export const createAudioOnlyEnableEvent = function() {
```
Comment (saghul) [inline react/features/analytics/AnalyticsEvents.js]: maybe we should combine the enabled / disabled functions into one, which takes a boolean instead.

File: react/features/app-settings/components/AppSettings.native.js [hunk 10]
```diff
@@ -56,110 +55,88 @@ class AppSettings extends AbstractAppSettings {
...
+                <View style = { PlatformElements.page }>
+                    <Header>
+                        <BackButton
+                            onPress = { this._onRequestClose } />
+                        <Text
+                            style = { [
+                                styles.text,
+                                PlatformElements.headerText
+                            ] } >
+                            { t('profileModal.header') }
```
Comment (saghul) [inline react/features/app-settings/components/AppSettings.native.js]: can we please rename all these xxxModal to something else? having "Modal" in the name doesn't really add anything and may become incorrect in the future.

File: react/features/app-settings/components/native/FormSectionHeader.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+class FormSectionHeader extends Component<Props> {
+    /**
+     * Implements React's {@link Component#render()}.
+     *
+     * @inheritdoc
+     * @override
+     * @returns {ReactElement}
+     */
+    render() {
+        const { t } = this.props;
```
Comment (saghul) [inline react/features/app-settings/components/native/FormSectionHeader.js]: please extract style and i18nlabel here too

File: react/features/app-settings/components/native/index.js [hunk 12]
```diff
@@ -0,0 +1,3 @@
+export { default as BackButton } from './BackButton';
+export { default as FormRow } from './FormRow';
+export { default as FormSectionHeader } from './FormSectionHeader';
```
Comment (saghul) [inline react/features/app-settings/components/native/index.js]: All these components looks pretty generic, aren't they? How about having them in features/mobile/forms or something similar, so we can eventually reuse them?

File: react/features/base/react/components/native/Header.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
+/* @flow */
```
Comment (saghul) [inline react/features/base/react/components/native/Header.js]: makr this `// @flow`, we are adding new code like that

File: react/features/base/react/components/native/Header.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * The type of the React {@code Component} props of {@link ScreenHeader}
+ */
+type Props = {
+
+    /**
+     * Children component(s).
+     */
+    children: Object,
```
Comment (saghul) [inline react/features/base/react/components/native/Header.js]: `React$Node` is more appropriate here

File: react/features/base/react/components/native/SideBar.js [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+
+    /**
+     * The local participant's avatar
+     */
+    _avatar: string,
+
+    /**
+     * The children of the Component
+     */
+    children: Object,
```
Comment (saghul) [inline react/features/base/react/components/native/SideBar.js]: ditto

File: react/features/base/react/components/native/SideBar.js [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+    Dimensions,
+    TouchableWithoutFeedback,
+    View
+} from 'react-native';
+
+import styles, { SIDEBAR_WIDTH } from './styles';
+
+
+/**
+ * The type of the React {@code Component} props of {@link AnimatedSideBar}
```
Comment (saghul) [inline react/features/base/react/components/native/SideBar.js]: s/Animated//g

File: react/features/base/react/components/native/SideBar.js [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+        super(props);
+
+        this.state = {
+            showSideBar: false,
+            showOverlay: false,
+            sliderAnimation: new Animated.Value(-SIDEBAR_WIDTH)
+        };
+
+        this._setShow = this._setShow.bind(this);
+
```
Comment (saghul) [inline react/features/base/react/components/native/SideBar.js]: why the newline?

File: react/features/base/react/components/native/SideBar.js [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+        this._setShow(props.show);
+    }
+
+    /**
+     * Implements the Component's componentDidMount method.
+     *
+     * @inheritdoc
+     */
+    componentDidMount() {
+        this._ismounted = true;
```
Comment (saghul) [inline react/features/base/react/components/native/SideBar.js]: we use `_mounted` elsewhere, let's be consistent

File: react/features/base/react/components/native/styles.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (saghul) [reply react/features/base/react/components/native/styles.js]: ack, can you please put a comment about it?

File: react/features/welcome/actions.js [hunk 19]
```diff
@@ -0,0 +1,17 @@
+import { SET_SIDEBAR_VISIBILITY } from './actionTypes';
+
+/**
+ * Redux action to hide or show the status bar.
+ *
+ * @param {boolean} visible - The new value of the visibility.
+ * @returns {{
+ *  type: SET_SIDEBAR_VISIBILITY,
```
Comment (saghul) [inline react/features/welcome/actions.js]: I think there is some spacing issue here, can you please look at other actions?

File: react/features/welcome/actions.js [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+ *  sideBarVisible: boolean
+ * }}
+ */
+export function setSideBarVisibility(visible) {
+    return {
+        type: SET_SIDEBAR_VISIBILITY,
+        sideBarVisible: visible
```
Comment (saghul) [inline react/features/welcome/actions.js]: better call this "visible", the fact that we are talking abut the sidebar is implied by the action type

File: react/features/welcome/actions.js [hunk 21]
```diff
@@ -0,0 +1,17 @@
+import { SET_SIDEBAR_VISIBILITY } from './actionTypes';
```
Comment (saghul) [inline react/features/welcome/actions.js]: pedantic, I know, but let's add flow here too

File: react/features/welcome/components/BlankPage.native.js [hunk 22]
```diff
@@ -54,7 +53,7 @@ class BlankPage extends Component<*> {
      */
     render() {
         return (
-            <LocalVideoTrackUnderlay style = { styles.blankPage }>
+            <LocalVideoTrackUnderlay>
```
Comment (saghul) [inline react/features/welcome/components/BlankPage.native.js]: Do we need a StatusBar component here?

File: react/features/welcome/components/WelcomePage.native.js [hunk 23]
```diff
@@ -65,49 +91,88 @@ class WelcomePage extends AbstractWelcomePage {
...
+
+    /**
+     * Handles the audio-video switch changes.
+     *
+     * @private
+     * @param {boolean} startWithVideo - The negaty of the new value
+     * of startAudioOnly.
+     * @returns {void}
+     */
+    _onStartAudioOnlyChange(startWithVideo) {
```
Comment (saghul) [inline react/features/welcome/components/WelcomePage.native.js]: Can you make this affirmative instead of negative? We have startWithVideoMuted, startAudioOnly, etc, and this my mix things up. 

Also, there is a double negation here, because you say this is the negated value, but then negate it again when updating the profile...

File: react/features/welcome/components/WelcomePage.native.js [hunk 24]
```diff
@@ -65,49 +91,88 @@ class WelcomePage extends AbstractWelcomePage {
...
+        const { dispatch } = this.props;
+
+        dispatch(updateProfile({
+            ...this.props._profile,
```
Comment (saghul) [inline react/features/welcome/components/WelcomePage.native.js]: Why is this needed? Can't we update the updateProfile action to only update the keys that are passed and keep the rest?

File: react/features/welcome/components/WelcomePage.native.js [hunk 25]
```diff
@@ -65,49 +91,88 @@ class WelcomePage extends AbstractWelcomePage {
...
+            startAudioOnly: !startWithVideo
+        }));
+
+        dispatch(setAudioOnly(!startWithVideo));
```
Comment (saghul) [inline react/features/welcome/components/WelcomePage.native.js]: I think this belongs in the profile middleware. This way, if / when we add the ability to change the profile mid-conference, it will Just Work.

File: react/features/welcome/components/WelcomePageSideBar.native.js [hunk 26]
```diff
@@ -0,0 +1,167 @@
+// @flow
+
+import React, { Component } from 'react';
+import { SafeAreaView, ScrollView, Text } from 'react-native';
+import { connect } from 'react-redux';
+
+import SideBarItem from './SideBarItem';
```
Comment (saghul) [inline react/features/welcome/components/WelcomePageSideBar.native.js]: Why is this component not together with the SideBar in base/react ?

File: react/features/welcome/components/WelcomePageSideBar.native.js [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+
+/**
+ * The URL at which the terms (of service/use) are available to the user.
+ */
+const TERMS_URL = 'https://jitsi.org/meet/terms';
+
+type Props = {
+
+    /**
+     * Redux display action
```
Comment (saghul) [inline react/features/welcome/components/WelcomePageSideBar.native.js]: s/display/dispatch/

File: react/features/welcome/components/WelcomePageSideBar.native.js [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+    _displayName: string,
+
+    /**
+     * Sets the side bar visible or hidden.
+     */
+    _visible: boolean
+};
+
+/**
+ * A component rendering a system sidebar.
```
Comment (saghul) [inline react/features/welcome/components/WelcomePageSideBar.native.js]: system?

File: react/features/welcome/components/styles.js [hunk 29]
```diff
@@ -17,87 +23,102 @@ const TEXT_COLOR = ColorPalette.white;
...
+
+    /**
+     * Style of the avatar in te side bar.
+     */
+    avatar: {
+        alignSelf: 'center',
+
+        // XXX Workaround for Android: for images < 80 the border radius doesn't
+        // work properly, but applying a radius twice as big does the trick.
+        borderRadius: Platform.OS === 'android' ? 200 : 50,
```
Comment (saghul) [inline react/features/welcome/components/styles.js]: I don't think you need this (I can test) since the avatar size is 100.

File: react/features/welcome/components/styles.js [hunk 30]
```diff
@@ -17,87 +23,102 @@ const TEXT_COLOR = ColorPalette.white;
...
+    itemContainer: {
+        flexDirection: 'column',
+        paddingTop: 10
     },
 
     /**
      * The style of the top-level container/{@code View} of
      * {@code LocalVideoTrackUnderlay}.
+     *
+     * fixAndroidViewClipping is removed as TintedView fixes the related issue.
```
Comment (saghul) [inline react/features/welcome/components/styles.js]: please put this in the commit log, there is no need to leave the comment here

File: react/features/welcome/reducer.js [hunk 31]
```diff
@@ -0,0 +1,20 @@
+import { ReducerRegistry } from '../base/redux';
+import { SET_SIDEBAR_VISIBILITY } from './actionTypes';
+
+/**
+ * Reduces the Redux actions of the feature features/recording.
+ */
+ReducerRegistry.register('features/welcome', (state = {
```
Comment (saghul) [inline react/features/welcome/reducer.js]: put each parameter in a new line

File: react/features/welcome/reducer.js [hunk 32]
```diff
@@ -0,0 +1,20 @@
...
+    sideBarVisible: false
```
Comment (saghul) [inline react/features/welcome/reducer.js]: make the default state a constant at the top level

File: react/features/welcome/reducer.js [hunk 33]
```diff
@@ -0,0 +1,20 @@
...
+}, action) => {
+    switch (action.type) {
+    case SET_SIDEBAR_VISIBILITY:
+        return {
+            ...state,
+            sideBarVisible: action.sideBarVisible
```
Comment (saghul) [inline react/features/welcome/reducer.js]: rename the action parameter to just "visible"

**Comment by saghul** [comment]
For the iOS < 11 status bar thing, maybe some of this comes useful: https://stackoverflow.com/a/39300715/280310

**Comment by zbettenbuk** [comment]
image

File: react/features/app-settings/components/native/index.js (hunk 12 above)
Comment (zbettenbuk) [reply react/features/app-settings/components/native/index.js]: I agree, but let's do this in another PR. I don't really want to further grow this PR, if not necessary. Then I'll address this, and the introduction of safe areas (your previous comment) on the conference screen in separate commits, if that's okay.

Also, @lyubomir what do you think?

File: react/features/base/react/components/native/SideBar.js (hunk 17 above)
Comment (zbettenbuk) [reply react/features/base/react/components/native/SideBar.js]: I like to separate logically disjoined blocks, like this one is a behaviour call, and the next 3 lines are function bindings. For me they don't really belong together

File: react/features/welcome/actions.js (hunk 20 above)
Comment (zbettenbuk) [reply react/features/welcome/actions.js]: Yeah, but this is the overall state of the WelcomeScreen, so a simple boolean 'visible' doesn't tell what is visible, when accessing the welcome screen state somewhere else.

File: react/features/welcome/components/BlankPage.native.js (hunk 22 above)
Comment (zbettenbuk) [reply react/features/welcome/components/BlankPage.native.js]: Hmm, not sure what you mean. I'd say not, but is there a header there?

File: react/features/welcome/components/WelcomePage.native.js (hunk 23 above)
Comment (zbettenbuk) [reply react/features/welcome/components/WelcomePage.native.js]: Yes. Also, this will be changed, because Emil asked to turn the switch the other direction: Video <> Voice

File: react/features/welcome/components/WelcomePage.native.js (hunk 24 above)
Comment (zbettenbuk) [reply react/features/welcome/components/WelcomePage.native.js]: Ok, but let me change this in the flatten redux tree PR instead, because these calls are affected by that change

File: react/features/welcome/components/WelcomePageSideBar.native.js (hunk 26 above)
Comment (zbettenbuk) [reply react/features/welcome/components/WelcomePageSideBar.native.js]: I don't feel it too generic. This is an item component that renders welcome page sidebar button-like components, while the SideBar component in base/react is so generic, that it's basically totally empty, and can even contain video, or anything.

File: react/features/welcome/reducer.js (hunk 33 above)
Comment (zbettenbuk) [reply react/features/welcome/reducer.js]: I explained this above

File: react/features/welcome/actions.js (hunk 20 above)
Comment (zbettenbuk) [reply react/features/welcome/actions.js]: E.g. it'll be not clear what is supposed to be visible when you see this in a _mapStateToProps call:
state['features/welcome'].visible

**Comment by zbettenbuk** [comment]
@jitsi/mobile finished fixing the PR, tests passed. Now that Saúl is off, can you please review? Thanks!
